core/use-swr.ts

```diff
@@ -183,6 +183,10 @@ export const useSWRHandler = <Data = any, Error = any>(
   )
 
   const isInitialMount = !initialMountedRef.current
+
+  const hasRevalidator =
+    EVENT_REVALIDATORS[key] && EVENT_REVALIDATORS[key].length > 0
```

Comment:
@promer94 Thank you for applying code review!

I reread the current implementation and couldn't understand why we need to check if there is another revalidator only when an error happened.

Isn't it enough to add a check the error is `undefined` like this?

```jsx
    // If there is no stale data and error, we need to revalidate when mount;
    // If `revalidateIfStale` is set to true, we will always revalidate.
    return isUndefined(error) && (isUndefined(data) || config.revalidateIfStale)
```

https://github.com/vercel/swr/blob/3cec4bd18a163bde31d8382915fcf0753d7194ed/core/use-swr.ts#L216-L218

Reply:
For most of time, we actully want to start revalidation after hooks mouted  when there is an error in cache.

But for edge case like #2336, we dont want second-mounted hook to start revalidation since the first-mouted hooks will handle it.


We just want a signle instantace to handle the revalidation process.  


https://github.com/vercel/swr/blob/06a644129e66982f2d837047be797df2e506b228/_internal/utils/cache.ts#L17-L24
https://github.com/vercel/swr/blob/3cec4bd18a163bde31d8382915fcf0753d7194ed/_internal/utils/mutate.ts#L106-L110

Reply:
@promer94 
Thank you! I understand the intention 👍 

Let me share another solution I came up with.
I feel the issue is caused by an incorrect `isLoading` state on an error, and I'd say `isLoading` should be `true` only when both `data` and `error` are `undefined`.

https://github.com/vercel/swr/blob/3cec4bd18a163bde31d8382915fcf0753d7194ed/core/use-swr.ts#L302-L304

should be

```jsx
      if (isUndefined(getCache().data) && isUndefined(getCache().error)) {
        initialState.isLoading = true
      }
```


What do you think?

Reply:
Good Catch !

But `isValidating` will also cause infinite retry here.  So I think we stiil need to check the number of hook to avoid it.

we could apply both changes and a extra test for ` isValidating`.

Reply:
> But isValidating will also cause infinite retry here. So I think we stiil need to check the number of hook to avoid it.

I missed the case, thank you!

Reply:
```jsx
 return isUndefined(error) && (isUndefined(data) || config.revalidateIfStale) 
```

This implmentation will fix the problem but it also will change the previous behavior. (start revalidation right after hooks mouted when there is an error in cache). It will not start revalidation immediately anymore. It has to wait for a focus event or auto revalidation.

> Should we also check the existing revalidators for non-error cases not only error cases to avoid revalidation by second-mounted hook? 

We actully did that for other kind of revalidators. 

https://github.com/vercel/swr/blob/06a644129e66982f2d837047be797df2e506b228/_internal/utils/cache.ts#L17-L24
https://github.com/vercel/swr/blob/3cec4bd18a163bde31d8382915fcf0753d7194ed/_internal/utils/mutate.ts#L106-L110


And for mounting revalidation, we also did same thing but in different way to prevent second mounted hook to start a new revalidation

https://github.com/vercel/swr/blob/c57edae96a0af03497f657027e2e784d2915965b/core/use-swr.ts#L258